landlab/utils/source_tracking_algorithm.py

```diff
@@ -0,0 +1,189 @@
+        # Check 1: Check if this node has been visited earlier. If yes,
+        # then skip to next node
+        if i in alr_counted:
+            continue
+        # Check 2: If the visited node is a sink
+        if r[i] == i:
+            hsd_upstr.update({i: [hsd_ids[i]]})
+            flow_accum[i] += 1.
+            alr_counted.append(i)
+            continue
+        # Check 3: Now, if the node is not a sink and hasn't been visited, it
+        # belongs to a stream segment. Hence, all the nodes in the stream will
+        # have to betraversed.
+        # stream_buffer is a list that will hold the upstream contributing
+        # node information for that particular segment until reaching outlet.
+        stream_buffer = []
+        j = i
+        switch_i = True
+        a = 0.
+        # Loop below will traverse the segment of the stream until an outlet
+        # is reached.
+        while True:
+            # Following if loop is to execute the contents once the first node
+            # in the segment is visited.
+            if not switch_i:
+                j = r[j]
+                if j not in core_nodes:
+                    break
+            # If this node is being visited for the first time,
+            # this 'if statement' will executed.
+            if flow_accum[j] == 0.:
+                a += 1.
+                alr_counted.append(j)
+                stream_buffer.append(hsd_ids[j])
+            # Update number of upstream nodes.
+            flow_accum[j] += a
+            # If the node is being visited for the first time, the dictionary
+            # 'hsd_upstr' will be updated.
+            if j in hsd_upstr.keys():
+                hsd_upstr[j] += copy.copy(stream_buffer)
+            # If the node has been already visited, then the upstream segment
+            # that was not accounted for in the main stem, would be added to
+            # all downstream nodes, one by one, until the outlet is reached.
+            else:
+                hsd_upstr.update({j: copy.copy(stream_buffer)})
+            # If the outlet is reached, the 'while' loop will be exited.
+            if r[j] == j:
+                break
+            # This will be executed only for the first node of the
+            # stream segment.
+            if switch_i:
+                switch_i = False
+    return (hsd_upstr, flow_accum)
+
+
+# %%
+# Algorithm to calculate coefficients of each upstream HSD ID
+def find_unique_upstream_hsd_ids_and_fractions(hsd_upstr):
+    """
+    Counts repeated HSD ids and maps MD nodes with unique upstream HSD ids
+    and corresponding fraction of contribution.
+    
+    Parameters:
+    ----------
+    hsd_upstr: dictionary
+        'hsd_upstr' maps each Model Domain (MD) grid node to corresponding
+        contributing upstream HSD ids.
+        
+    Returns:
+    -------
+    (unique_ids, fractions): (dictionary, dictionary)
+        Tuple of data. 'unique_ids' maps each MD node with all upstream HSD
```

Comment:
Well, we have used MD and HSD for Landslides User Manual and we would like to be consistent with these names. I have tried to emphasize that the grid referred here is a Landlab RasterModelGrid instance though.